### PR TITLE
[pytorch] bump up variable version regardless of differentiability

### PR DESCRIPTION
**Pull request by ljk53**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#41269 [pytorch] bump up variable version regardless of differentiability**

The ultimate goal is to move things that are not gated with `if (compute_requires_grad(...))`
or `if (grad_fn)` out from VariableType so that VariableType kernels can be enabled/disabled
based upon `GradMode`. Then we can merge `AutoNonVariableTypeMode` and `NoGradGuard`.

We've moved profiling / tracing logic out from VariableType. One remaining thing that's
not gated with the if-statement is the `increment_version` call.

However, the `gen_variable_type.py` does use bits from `derivatives.yaml` to determine whether
to emit the `increment_version` call. If an output is never going to be differentiable (not based
upon runtime property of the variable but based upon static property, e.g. it's integral type)
then it would never emit the increment_version call.

Hypothetically, increment_version for a tensor can be orthogonal to its differentiability.

This PR is to make the change and test its impact. Making this logical simplification would
allow us to move this out from VariableType to aten codegen.

Differential Revision: [D22471643](https://our.internmc.facebook.com/intern/diff/D22471643/)

Differential Revision: [D22471643](https://our.internmc.facebook.com/intern/diff/D22471643)

### CODE REVIEW

File: tools/autograd/gen_variable_type.py [hunk 1]
```diff
@@ -1206,10 +1206,11 @@ def emit_increment_version():
     body.append(declare_returned_variables)
 
     body.append(emit_call(env, tie_return_values))
+    if strategy == 'use_derived':
+        body.extend(emit_increment_version())
```
Comment (ezyang) [inline tools/autograd/gen_variable_type.py:1210]: So the idea here is, if the op in question is composite, we DON'T emit the increment version (so that the constituent pieces can take care of it)

File: tools/autograd/gen_variable_type.py [hunk 2]
```diff
@@ -1189,7 +1189,7 @@ def emit_check_inplace():
     def emit_increment_version():
         if not modifies_arguments:
             return []
-        return ['increment_version({});'.format(arg['name']) for arg in differentiable_outputs]
+        return ['increment_version({});'.format(arg['name']) for arg in returns]
```
Comment (albanD) [inline tools/autograd/gen_variable_type.py:1192]: Why is this changed?  does it actually matter?
Do we have inplace ops that return multiple things? And if so do we have some that return a mix of differentiable/non-differentiable outputs?

**Comment by albanD** [comment]
> The only remaining thing that's not gated with the if-statement is the increment_version call.

What about the view handling via the `view_as()` calls? These are not generated for non-differentiable functions either but should still be executed when grad mode is disabled.

> Hypothetically, increment_version for a tensor can be orthogonal to its differentiability.

But the differentiability impact the way we call the function (composite or not) and the way we call the function impact if we call increment_version or not.
So they are not really orthogonal?

File: tools/autograd/gen_variable_type.py (hunk 2 above)
Comment (ezyang) [reply tools/autograd/gen_variable_type.py:1192]: We should increment version on all tensors that get mutated, not just the differentiable ones. You can save non differentiable tensors as part of backwards formula...

**Comment by ezyang** [comment]
> What about the view handling via the view_as() calls? These are not generated for non-differentiable functions either but should still be executed when grad mode is disabled.

Hm, yes, this is probably more long tail stuff we will have to handle.

**Review by ezyang — APPROVED** [review]

File: tools/autograd/gen_variable_type.py (hunk 2 above)
Comment (albanD) [reply tools/autograd/gen_variable_type.py:1192]: But I don't expect that these other outputs are actually modified inplace! We would be bumping the version of a Tensor we don't change inplace.

Also this is BC-breaking right?

Comment (ljk53) [reply tools/autograd/gen_variable_type.py:1192]: > Why is this changed? does it actually matter?
> Do we have inplace ops that return multiple things? And if so do we have some that return a mix of differentiable/non-differentiable outputs?

There are cases where it returns multiple things mixed of differentiable/non-differentiable outputs, e.g.:
https://github.com/ljk53/pytorch/commit/9ab9e4bc3489c9b3391bed174305d7a6b00457b5#diff-79b1a31c97eee8dda9e0dae02162beecR2986

non-differentiable outputs are usually things like `indices`.

Comment (albanD) [reply tools/autograd/gen_variable_type.py:1192]: Ho right these `out=(val, ind)` functions...
Ok!

**Review by albanD — APPROVED** [review]
Sounds good.

Please do add a note about BC-breaking changes this leads to in the main comment so that we can add it to the release doc.

**Comment by ljk53** [comment]
> Sounds good.
> 
> Please do add a note about BC-breaking changes this leads to in the main comment so that we can add it to the release doc.

Thanks for reviewing and approving this PR! I was actually using this PR as an experiment vehicle to trigger CIs so I didn't add any reviewers, guess I should have marked it as WIP, lol...

I might still experiment something else before landing it.

File: tools/autograd/gen_variable_type.py (hunk 2 above)
Comment (ljk53) [reply tools/autograd/gen_variable_type.py:1192]: > But I don't expect that these other outputs are actually modified inplace! We would be bumping the version of a Tensor we don't change inplace.
> 
> Also this is BC-breaking right?

This is a good point - just to clarify: regarding BC breakage, do you mean:
1. code that uses `x._version` to read variable version explicitly, or
2. code broken by falsely incrementing version for tensors that are not actually updated?

I call out "falsely" bumping because it's harmful as the breakage is totally unnecessary compared to truly bumping - in which case it is technically more correct behavior - if truly bumping breaks any code then it probably reveals bugs.

Seems we made effort to differentiate `Tensor&` and `const Tensor&` and seems `returns` at this place only contains `Tensor&` ones - so I assumed that these params are possibly mutated and bumped their versions. By eyeballing check seems in most cases the `returns` are indeed mutated, but I haven't verified all cases - do you know whether is it right assumption?

Comment (albanD) [reply tools/autograd/gen_variable_type.py:1192]: > code that uses x._version to read variable version explicitly

This one is ok to break as this is an internal API. You might need to update a couple tests but it should be fine overall (even though we want to avoid breaking it often).

> code broken by falsely incrementing version for tensors that are not actually updated?

There are two cases here indeed:
- Cases where we were not incrementing while we should:
```python
import torch
from torch.utils import checkpoint

a = torch.ones(10, 10, requires_grad=True)

b, ind = a.max(dim=0)

with torch.no_grad():
    if False:
        ind += 1 # Raise an error as expected
    elif True:
        # No error and wrong grad
        t = torch.zeros(10)
        t[2] = 1
        torch.cummax(t, dim=0, out=(torch.Tensor(), ind))
    else:
        pass

b.sum().backward()
print(a.grad)
```
- Cases where we should not increment as the second result is not modified inplace. Not sure if this happens in practice. In any case, we should be able to tell from the signature if it is modified or not (the signature is (should) always right!).

**Comment by albanD** [comment]
Thinking more about it, I would agree that increment should not be associated with the differentiability of a function.
Because even if a function is not relevant for gradient computation, it should not lead to silently wrong gradients by side effect.
